unified_planning/grpc/proto_writer.py

```diff
@@ -0,0 +1,534 @@
+# Copyright 2021 AIPlan4EU project
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import fractions
+
+import unified_planning.grpc.generated.unified_planning_pb2 as unified_planning_pb2
+import unified_planning.model
+import unified_planning.plan
+from unified_planning import model
+from unified_planning.exceptions import UPException
+from unified_planning.grpc.converter import Converter, handles
+from unified_planning.model.operators import (
+    BOOL_OPERATORS,
+    IRA_OPERATORS,
+    RELATIONS,
+    OperatorKind,
+)
+from unified_planning.model.timing import TimepointKind
+
+
+def map_operator(op: int) -> str:
+    if op == OperatorKind.PLUS:
+        return "+"
+    elif op == OperatorKind.MINUS:
+        return "-"
+    elif op == OperatorKind.TIMES:
+        return "*"
+    elif op == OperatorKind.DIV:
+        return "/"
+    elif op == OperatorKind.LE:
+        return "<="
+    elif op == OperatorKind.LT:
+        return "<"
+    elif op == OperatorKind.EQUALS:
+        return "="
+    elif op == OperatorKind.AND:
+        return "and"
+    elif op == OperatorKind.OR:
+        return "or"
+    elif op == OperatorKind.NOT:
+        return "not"
+    elif op == OperatorKind.IMPLIES:
+        return "implies"
+    elif op == OperatorKind.IFF:
+        return "iff"
+    elif op == OperatorKind.EXISTS:
+        return "exists"
+    elif op == OperatorKind.FORALL:
+        return "forall"
+
+    raise ValueError(f"Unknown operator `{op}`")
+
+
+def map_feature(feature: str):
+    pb_feature = unified_planning_pb2.Feature.Value(feature)
+    if pb_feature is None:
+        raise ValueError(f"Cannot convert feature to protobuf {feature}")
+    return pb_feature
+
+
+class ProtobufWriter(Converter):
+    @handles(unified_planning.model.Fluent)
+    def _convert_fluent(self, fluent, problem: model.Problem):
+        name = fluent.name
+        sig = [self.convert(t) for t in fluent.signature]
+        return unified_planning_pb2.Fluent(
+            name=name,
+            value_type=str(fluent.type),
+            parameters=sig,
+            default_value=self.convert(problem.fluents_defaults[fluent])
+            if fluent in problem.fluents_defaults
+            else None,
+        )
+
+    @handles(unified_planning.model.Object)
+    def _convert_object(self, obj):
+        return unified_planning_pb2.ObjectDeclaration(name=obj.name, type=str(obj.type))
+
+    @handles(unified_planning.model.FNode)
+    def _convert_fnode(self, exp):
```

Comment:
Here we could also use a `walker` to convert the FNode

Reply:
@mikand what do you think here, about using a walker for the FNode conversion?
Anyhow the current implementation works.